### ISSUE TITLE
How to replace a parameter with other variable while keeping the backpropagation?

### ISSUE THREAD

**Issue by kunwuz** [pytorch/pytorch#40165]
For example, now I have a parameter 'D' in the model.

Now I want to replace the 'D' with 'C', where 'C = a+b'. Is there anyway in pytorch that can achieve that replacement while keeping the backpropagation between 'C' and 'a+b'. (e.g., training the model will update the value of 'a' and 'b'.

I've tried D.data = C, but obviously that just changed the value and violate the backpropagation. Besides, '.copy' and '.clone' didn't work either.

**Comment by albanD** [comment]
Hi,

We use github issues only for bugs or feature requests.
Please use the forum to ask questions: https://discuss.pytorch.org/

For your question, it is not possible to have a Parameter with some history. Since a Parameter is a leaf for which .backward() computes gradient. If it has history, it is not a leaf anymore and so it won't work.
If you want to have an attribute `self.foo` on your nn.Module that has history (it will be a regular Tensor, not a Parameter). This is possible. Note that if `self.foo` used to be a Parameter, you will need to first delete the parameter: `del self.foo` then you will be able to set a Tensor with history there with `self.foo = a+b`.

I'm going to close this as this is not a bug/feature.
Feel free to open a thread on the forum if you have follow up questions!
